### ISSUE TITLE
Single-color backlight keyboard: main window does not open

### ISSUE THREAD

**Issue by rotgruengelb** [OmenMon/OmenMon#25]
## Microsoft .NET Framework Exception when trying to access/click on "Show Monitor"
 
* OmenMon Version: **0.59.0-Release**
* Model & Version: **Victus by HP 16-d0xxx** / **HP 88F9 88.58**
* Windows: **23H2**
* Working:
  * Fan Control
* Not Working:
  * Monitor GUI

```
System.IndexOutOfRangeException: Der Index war außerhalb des Arraybereichs.
   bei OmenMon.AppGui.GuiKbd.SetColors(ColorTable colorTable)
   bei OmenMon.AppGui.GuiKbd.GetHw()
   bei OmenMon.AppGui.GuiKbd..ctor(GuiTray context)
   bei OmenMon.AppGui.GuiFormMain..ctor()
   bei OmenMon.AppGui.GuiTray.ShowFormMain()
   bei OmenMon.AppGui.GuiTray.EventIconMouseClick(Object sender, MouseEventArgs e)
   bei System.Windows.Forms.NotifyIcon.OnMouseClick(MouseEventArgs mea)
   bei System.Windows.Forms.NotifyIcon.WmMouseUp(Message& m, MouseButtons button)
   bei System.Windows.Forms.NotifyIcon.WndProc(Message& msg)
   bei System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
```

**Comment by GeographicCone** [comment]
Thank you for your interest in **OmenMon** and the report.

This is related to a possibly unsupported keyboard not detected as such.

Can you please say what kind of keyboard backlight type do you have and post your `OmenMon -Bios` output (or just `OmenMon -Bios KbdType HasBacklight Backlight Color` specifically)?

**Comment by rotgruengelb** [comment]
This is the output of `OmenMon -Bios KbdType HasBacklight Backlight Color`:

```
OmenMon Hardware Monitoring & Control Utility Version 0.59-Release
BIOS (-Bios)
- Keyboard Type: 0x00 = 0b00000000 = 0 [Standard]
- Keyboard Backlight Support: Yes
- Keyboard Backlight: 0x64 = 0b01100100 = 100 [Off]
- Keyboard Backlight Color Table: 1 Zones - Right: 000000
```

**Comment by GeographicCone** [comment]
Great, thank you!

So with your keyboard, are you able to set the backlight to an arbitrary RGB color?

Or is it a single, predetermined color (likely either red or white), which you can only toggle on and off?

Meanwhile, before we get this issue sorted out, you should be able to use the rest of **OmenMon**'s functionality by temporarily setting [BiosErrorReporting](https://omenmon.github.io/config#bioserrorreporting) to **false** in `OmenMon.xml`.

**Comment by rotgruengelb** [comment]
> Or is it a single, predetermined color (likely either red or white), which you can only toggle on and off?

Yes Im only able toggle a white key backlight on or off

>Meanwhile, before we get this issue sorted out, you should be able to use the rest of **OmenMon**'s functionality by temporarily setting [BiosErrorReporting](https://omenmon.github.io/config#bioserrorreporting) to **false** in `OmenMon.xml`.

The bios reporting false was already done but did nothing

**Comment by GeographicCone** [comment]
Could you please try the test build from this artifact and see if the issue is resolved for you:
https://github.com/OmenMon/OmenMon/actions/runs/7449156009

You should be able to toggle the backlight on and off from the main window or the context menu but the color-related part of the interface should be grayed out and there should be no errors of any kind of course.

**Comment by rotgruengelb** [comment]
## Working
- Monitor Opening.

## Not Working
- "Keyboard Backlight & Color" Section Check Mark does nothing when clicked.
- Keyboard > Backlight Dose nothing (Keyboard Backlight doesn't change)


## Test Video

Uploading 2024-01-09 14-31-58.mp4…

**Comment by GeographicCone** [comment]
Thank you!
One more thing:

Does the backlight toggle work when you run:
`OmenMon -Bios Backlight=On`
and
`OmenMon -Bios Backlight=Off`
from the command line?

**Comment by rotgruengelb** [comment]
```
C:\Users\Daniel\Desktop\OmenMon\experimental>OmenMon -Bios Backlight=Off
OmenMon Hardware Monitoring & Control Utility Version 0.60-Snapshot
BIOS (-Bios)
+ Keyboard Backlight: No
C:\Users\Daniel\Desktop\OmenMon\experimental>OmenMon -Bios Backlight=On
OmenMon Hardware Monitoring & Control Utility Version 0.60-Snapshot
BIOS (-Bios)
+ Keyboard Backlight: Yes
C:\Users\Daniel\Desktop\OmenMon\experimental>OmenMon -Bios Backlight=Off
OmenMon Hardware Monitoring & Control Utility Version 0.60-Snapshot
BIOS (-Bios)
+ Keyboard Backlight: No
```
The backlight did not change for any command

**Comment by DilutedOxygen1451** [comment]
> Could you please try the test build from this artifact and see if the issue is resolved for you: https://github.com/OmenMon/OmenMon/actions/runs/7449156009
> 
> You should be able to toggle the backlight on and off from the main window or the context menu but the color-related part of the interface should be grayed out and there should be no errors of any kind of course.

i attempted to build the version  linked, however on running the -prepare, it failed, telling me this issue
![image](https://github.com/OmenMon/OmenMon/assets/95299691/4c42f329-af00-40c8-9860-fedffcd0dc57)

**Comment by DilutedOxygen1451** [comment]
> Could you please try the test build from this artifact and see if the issue is resolved for you: https://github.com/OmenMon/OmenMon/actions/runs/7449156009
> 
> You should be able to toggle the backlight on and off from the main window or the context menu but the color-related part of the interface should be grayed out and there should be no errors of any kind of course.

running "make prepare" results in this error within the main folder results in this

![image](https://github.com/OmenMon/OmenMon/assets/95299691/36978086-41ab-4bbe-907a-5bc2e33076c6)


 and attempting to run (make build) results in an error telling me "OmenMon.exe could not be found"

**Comment by GeographicCone** [comment]
@rotgruengelb

> The backlight did not change for any command

Thank you for reporting back. So the backlight toggle WMI calls that seem to work on most *Omen*s are not compatible with that _Victus_ model.

On a general note, I'm really happy **OmenMon** has been found operational, at least partially, on such a variety of laptops, even including some _Envy_ (i.e. non-gaming) models and I always welcome any ideas how to extend the compatibility even further.

As a last resort, even if it doesn't work, **OmenMon** should always fail gracefully, and not with some ugly exception: I'm always willing to address issues like that. Beyond that though, in this case I don't know of any other way for toggling keyboard backlight programmatically, and considering it's likely not a priority issue since there must be some <key>Fn</key> key combination to do just that, I'm afraid I'll have to leave things the way they are for now. I appreciate your understanding.

@DilutedOxygen1451

> I attempted to build the version linked [...]

Apologies for being unclear. It's already been built by GitHub actions, just download the ZIP build artifact, unpack it, and run the `.exe` from there: https://github.com/OmenMon/OmenMon/actions/runs/7449156009/artifacts/1154269868

For the record, if you wanted to build **OmenMon** yourself, you'd have to install _Build Tools_ or _Visual Studio_ version 2022 (17). The details how to do that are available here: https://omenmon.github.io/build#instructions

But I'd never ask or expect anyone to run a build themselves just because they reported an issue. Sorry for the hassle I caused you due to this misunderstanding.

**Thank you both for for giving OmenMon a try!**

The linked preview will become the next release if no issues are found with the fix.

**Comment by DilutedOxygen1451** [comment]
ah, I downloaded it and it works flawlessly so far, I haven't tested toggling the backlight because fn+f4 on my keyboard does exactly that so i have no need, but the GUI is working perfectly as far as i can tell, thanks for the help

also its not your fault i misunderstood, im still relatively new to github and from the original artifact link i only found the unbuilt version, as i dont exactly know how to navigate github as much as i should, but the link you subsequently sent works perfectly

**Comment by rotgruengelb** [comment]
> As a last resort, even if it doesn't work, **OmenMon** should always fail gracefully, and not with some ugly exception: I'm always willing to address issues like that. Beyond that though, in this case I don't know of any other way for toggling keyboard backlight programmaticall,[...]

Thank you!
I never asked for backlight support and don't need it because there is indeed a <key>fn</key> key combination.

I just needed OmenMon for the fan control and it bothered me how it crashed every time I accedentaly clicked the taskbar icon that opens 
the monitor. 

Consider this issue resolved (at least for me).

Thank you @GeographicCone for your OpenSource contributions to this alternative to the "HP OmenHub garbage".

<3 -Daniel

**Comment by DilutedOxygen1451** [comment]
the issue is resolved for me too, thanks for the help, and I'm glad i can finally get rid of HP's garbage software, at least as much as i can

**Comment by GeographicCone** [comment]
Thank you both!
The fix is included in **build 0.60.0**, which [has just been released](https://github.com/OmenMon/OmenMon/releases/latest).
